source/extensions/tracers/xray/tracer.cc

```diff
@@ -1,37 +1,133 @@
+// This includes:
+//
+// - The version number, that is, 1.
+// - The time of the original request, in Unix epoch time, in 8 hexadecimal digits.
+// - A 96-bit unique identifier in 24 hexadecimal digits.
+//
+// For more details see:
+// https://docs.aws.amazon.com/xray/latest/devguide/xray-api-segmentdocuments.html#api-segmentdocuments-fields
+std::string generateTraceId(Envoy::SystemTime point_in_time) {
+  using namespace std::chrono;
+  Runtime::RandomGeneratorImpl rng;
+  const auto epoch = time_point_cast<seconds>(point_in_time).time_since_epoch().count();
+  std::string out;
+  out.reserve(35);
+  out += XRaySerializationVersion;
+  out.push_back('-');
+  // epoch in seconds represented as 8 hexadecimal characters
+  out += Hex::uint32ToHex(epoch);
+  out.push_back('-');
+  std::string uuid = rng.uuid();
+  // unique id represented as 24 hexadecimal digits and no dashes
+  uuid.erase(std::remove(uuid.begin(), uuid.end(), '-'), uuid.end());
+  ASSERT(uuid.length() >= 24);
+  out += uuid.substr(uuid.length() - 24);
+  return out;
+}
+
+} // namespace
+
+void Span::setId(uint64_t id) { id_ = Hex::uint64ToHex(id); }
+
 void Span::finishSpan() {
-  end_time_ = time_source_.systemTime();
-  // TODO(marcomagdy): serialize this span and send it to X-Ray Daemon
+  using std::chrono::time_point_cast;
+  using namespace envoy::tracers::xray;
+  // X-Ray expects timestamps to be in epoch seconds with milli/micro-second precision as a fraction
+  using SecondsWithFraction = std::chrono::duration<double>;
+  if (!sampled()) {
+    return;
+  }
+
+  daemon::Segment s;
+  s.set_name(name());
+  s.set_id(Id());
+  s.set_trace_id(traceId());
+  s.set_start_time(time_point_cast<SecondsWithFraction>(startTime()).time_since_epoch().count());
+  s.set_end_time(
+      time_point_cast<SecondsWithFraction>(time_source_.systemTime()).time_since_epoch().count());
+  s.set_parent_id(parentId());
+  using KeyValue = Protobuf::Map<std::string, std::string>::value_type;
+  for (const auto& item : custom_annotations_) {
+    s.mutable_annotations()->insert(KeyValue{item.first, item.second});
+  }
+
+  for (const auto& item : http_request_annotations_) {
+    s.mutable_http()->mutable_request()->insert(KeyValue{item.first, item.second});
+  }
+
+  for (const auto& item : http_response_annotations_) {
+    s.mutable_http()->mutable_response()->insert(KeyValue{item.first, item.second});
+  }
+
+  Protobuf::util::JsonPrintOptions json_options;
+  json_options.preserve_proto_field_names = true;
+  std::string json;
+  const auto status = Protobuf::util::MessageToJsonString(s, &json, json_options);
+  if (!status.ok()) {
+    throw new EnvoyException("Failed to serialize span to JSON.");
+  }
+
+  broker_->send(json);
 }
 
 void Span::injectContext(Http::HeaderMap& request_headers) {
-  const std::string xray_header_value =
-      fmt::format("root={};parent={};sampled=1", traceId(), parent_segment_id_);
+  const std::string& xray_header_value =
```

Comment:
why reference?

Reply:
It's a const reference, so it extends the lifetime of the temporary. It's cheaper than a move, unless the the compiler does (N)RVO, but I double fmt::format() is simple enough for (N)RVO to kick in.

Reply:
went ahead and wrote some tests to see how many allocations fmt does and to my pleasant surprise, it does a single allocation. so I'll remove the reference.
Thanks.

Reply:
even though that's unrelated to (N)RVO. But it seems like fmt is pretty darn optimized.